src/mlpack/methods/isomap/isomap.hpp

```diff
@@ -0,0 +1,129 @@
+ * variations of constructing neighbourhood graphs and finding all pair
+ * shortest paths in the graph. Presently, neighbourhood graph is constructed
+ * using only K_Nearest_Neighbours and all pair shortest path is calcuated
+ * using only Dijkstra's Algorithm
+ * 
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ISOMAP_ISOMAP_HPP
+#define MLPACK_METHODS_ISOMAP_ISOMAP_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/metrics/lmetric.hpp>
+#include "k_nearest.hpp"
+#include "dijkstra.hpp"
+#include "mds.hpp"
+
+namespace mlpack {
+namespace isomap {
+
+/**
+ * This class implements the Isomap algorithm which is part of Manifold
+ * Learning algorithms, and is used for dimensionality reduction. This
+ * algorithm converts the dataset into a lower dimensional space which
+ * it is assumed to belong from the higher dimensional space it is
+ * currently embedded into.
+ * This implementation is based on the paper:
+ * Lawrence Cayton, 2005, 17, Algorithms for Manifold Learning
+ */
+
+template <typename NeighbourhoodFunction = K_Nearest,
+          typename ShortestPathAlgo = Dijkstra>
+class Isomap
+{
+ public:
+  /**
+   * Create the Isomap obejct and set the n_neighbours parameter to
+   * specify the number of neighbours to consider while constructing
+   * neighbourhood graph.
+   * 
+   * @param n_neighbours -number of neighbours to consider
+   *                      for neighbourhood graph
+   */
+
+  Isomap(const size_t n_neighbours,
+         const NeighbourhoodFunction& neighbourhood = NeighbourhoodFunction(),
+         const ShortestPathAlgo& shortestPath = ShortestPathAlgo()) :
+         n_neighbours(n_neighbours),
+         neighbourhood(neighbourhood),
+         shortestPath(shortestPath)
+  { }
+
+  /**
+   * This is the main driver function to perform Isomap. Just the input
+   * matrix is required and it is safe to pass it as reference.
+   * 
+   * @param input -the input dataset to perform Isomap on.
+   */
+  void Apply(arma::mat& input)
+  {
+    arma::mat disMat(input.n_cols, input.n_cols);
+
+    // calculating distance matrix from the given input matrix
+    CalcDistanceMatrix(input, disMat);
+
+    // constructing neighbourhood graph (K_nearest is used)
+    neighbourhood.MakeNeighbourhoodGraph(disMat, n_neighbours);
+
+    // finding all pair shortest path in the neighbourhood graph created
+    shortestPath.FindShortestPath(disMat);
+
+    // making the shortest distance matrix symmetric (required for cMDS)
+    disMat = arma::min(disMat, disMat.t());
+
+    // mds object to perform classical multidimensional scaling
+    MDS md;
+
+    // performing classical multidimensional scaling
+    md.Apply(disMat);
```

Comment:
The API of overwriting the distance matrix (which here is not the distance matrix but the geodesic distance matrix or dissimilarity matrix, right?) is a little bit weird.  Why not provide a different matrix for the output, like, e.g., `md.Apply(disMat, output)`?

Reply:
I wanted to have result stored in the same space. But I don't think that will make much difference. And this is definitely better for understanding and reading. Also, since we are making a separate mds/ for mds, including an output is better.
Next commit will resolve this.